### PR TITLE
DDL-style no-csc2 schema change commands do not work on time partitions

### PR DESCRIPTION
**Pull request by nirbhayc**
Fixed by retrieving the schema for the time partition from its latest shard.

### CODE REVIEW

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/96169d5/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/96169d5/robomark.txt'>Error &#x26A0; </a>.<br>Regression testing: <b>4/371 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>cldeadlock [core dumped]<br>systbl_proc_leak<br>autoanalyze<br>sc_swapfields_logicalsc_generated</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/96169d5/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/96169d5/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2490'>Resubmit</a></p>

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/4e406a6/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: Success &#10003;.</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/4e406a6/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/4e406a6/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2490'>Resubmit</a></p>

File: db/views.c [hunk 1]
```diff
@@ -2519,13 +2519,15 @@ void views_unlock(void)
...
-char *timepart_newest_shard(const char *view_name, unsigned long long *version)
+char *timepart_newest_shard(const char *view_name, unsigned long long *version,
+                            int lock)
 {
     timepart_views_t *views = thedb->timepart_views;
     timepart_view_t *view;
     char *ret = NULL;
 
-    Pthread_rwlock_rdlock(&views_lk);
+    if (lock)
```
Comment (dorinhogea) [inline db/views.c:2529]: Is there a problem with nested read-locks?

Comment (nirbhayc) [reply db/views.c:2529]: No, but read somewhere that it isn't considered a good practice.

**Comment by dorinhogea** [comment]
I would suggest sticking with sqlite_master, to avoid rolling out issues.  Tools and cross-db sql will break.

File: db/views.c (hunk 1 above)
Comment (dorinhogea) [reply db/views.c:2529]: I find the extra code no needed

Comment (nirbhayc) [reply db/views.c:2529]: I think it's needed. If we allow timepart_newest_shard() unlock the rdlock, I'll have to acquire it again in retrieve_schema().

**Review by markhannum — APPROVED** [review]

**Comment by markhannum** [comment]
LGTM!

File: sqlite/src/comdb2build.c [hunk 2]
```diff
@@ -4258,6 +4285,8 @@ static int retrieve_schema(Parse *pParse, struct comdb2_ddl_context *ctx)
                 listc_init(&new_tag->column_list,
                            offsetof(struct comdb2_column, lnk));
                 if (retrieve_columns(pParse, ctx, old_tag, new_tag)) {
+                    if (views_lk_acquired)
```
Comment (dorinhogea) [inline sqlite/src/comdb2build.c:4288]: lets keep taglock inside, and view locks outside; 
goto err; here?

File: sqlite/src/comdb2build.c [hunk 3]
```diff
@@ -4090,6 +4115,8 @@ static int retrieve_schema(Parse *pParse, struct comdb2_ddl_context *ctx)
 
     /* Retrieve table columns. */
     if (retrieve_columns(pParse, ctx, schema, ctx->schema)) {
+        if (views_lk_acquired)
```
Comment (dorinhogea) [inline sqlite/src/comdb2build.c:4118]: goto err;?

File: sqlite/src/comdb2build.c [hunk 4]
```diff
@@ -4074,13 +4076,36 @@ static int retrieve_schema(Parse *pParse, struct comdb2_ddl_context *ctx)
...
+    } else {
+        Pthread_rwlock_unlock(&views_lk);
+        views_lk_acquired = 0;
+        table = get_dbtable_by_name(ctx->schema->name);
+    }
+
     if (table == 0) {
         pParse->rc = SQLITE_ERROR;
         sqlite3ErrorMsg(pParse, "Table '%s' not found.", ctx->schema->name);
+        if (views_lk_acquired)
```
Comment (dorinhogea) [inline sqlite/src/comdb2build.c:4107]: goto err;?

File: sqlite/src/comdb2build.c [hunk 5]
```diff
@@ -4074,13 +4076,36 @@ static int retrieve_schema(Parse *pParse, struct comdb2_ddl_context *ctx)
...
+    if (timepart_is_timepart(ctx->schema->name, 0)) {
+        unsigned long long version;
+        char *viewname;
+
+        viewname = timepart_newest_shard(ctx->schema->name, &version);
+        if (!viewname) {
+            sqlite3ErrorMsg(pParse,
+                            "Failed to retrieve shard information about '%s'",
+                            ctx->schema->name);
+            Pthread_rwlock_unlock(&views_lk);
```
Comment (dorinhogea) [inline sqlite/src/comdb2build.c:4094]: goto err;?

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>1/372 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>logfill</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/51327f9/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/51327f9/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2490'>Resubmit</a></p>

**Review by roborivers — APPROVED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: Success &#10003;.</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/08faadc/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/08faadc/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2490'>Resubmit</a></p>

**Comment by nirbhayc** [comment]
@markhannum @dorinhogea Thanks for the review!

**Review by dorinhogea — COMMENTED** [review]
LGTM, thank you

**Pull request merged**